python/spark_sklearn/tests/test_keyed_models.py

```diff
@@ -0,0 +1,237 @@
+
+        nExamplesPerUser = lambda i: max(minExamples, i + 1)
+        userKeys = [[i for _ in keyCols] for i in range(NUSERS)]
+        features = [[featureGen() for _ in range(nExamplesPerUser(i))] for i in range(NUSERS)]
+        useless = [["useless col" for _ in range(nExamplesPerUser(i))] for i in range(NUSERS)]
+        if isTransformer:
+            labels = None
+        else:
+            labels = [[labelGen() for _ in range(nExamplesPerUser(i))] for i in range(NUSERS)]
+
+        Xs = [np.vstack(x) for x in features]
+        ys = repeat(None) if isTransformer else [np.array(y) for y in labels]
+        localEstimators = [sklearn.base.clone(kwargs["sklearnEstimator"]).fit(X, y)
+                           for X, y in zip(Xs, ys)]
+        expectedDF = pd.DataFrame(userKeys, columns=keyCols)
+        expectedDF["estimator"] = localEstimators
+
+        def flattenAndConvertNumpy(x):
+            return [Vectors.dense(i) if isinstance(i, np.ndarray) else i
+                    for i in chain.from_iterable(x)]
+
+        inputDF = pd.DataFrame.from_dict(
+            {k: [i for i in range(NUSERS) for _ in range(nExamplesPerUser(i))] for k in keyCols})
+        inputDF[xCol] = flattenAndConvertNumpy(features)
+        inputDF["useless"] = flattenAndConvertNumpy(useless)
+        if labels: inputDF[kwargs["yCol"]] = flattenAndConvertNumpy(labels)
+        inputDF = self.spark.createDataFrame(inputDF)
+
+        ke = KeyedEstimator(**kwargs)
+        km = ke.fit(inputDF)
+
+        actualDF = km.keyedModels.toPandas()
+        _assertPandasAlmostEqual(actualDF, expectedDF, keyCols)
+
+        # Test users with different amounts of points.
+        nTestPerUser = lambda i: NUSERS // 4 if i < NUSERS // 2 else NUSERS * 3 // 4
+        testFeatures = [[featureGen() for _ in range(nTestPerUser(i))] for i in range(NUSERS)]
+        # "useless" column has nothing to do with computation, but is essential for keeping order
+        # the same between the spark and non-spark versions
+        useless = [range(nTestPerUser(i)) for i in range(NUSERS)]
+        inputDF = pd.DataFrame.from_dict(
+            {k: [i for i in range(NUSERS) for _ in range(nTestPerUser(i))] for k in keyCols})
+        inputDF[xCol] = flattenAndConvertNumpy(testFeatures)
+        inputDF["useless"] = flattenAndConvertNumpy(useless)
+
+        def makeOutput(estimator, X):
+            if isTransformer:
+                return estimator.transform(X)
+            else:
+                return estimator.predict(X).tolist()
+        Xs = [np.vstack(x) for x in testFeatures]
+        expectedOutput = map(makeOutput, localEstimators, Xs)
+        expectedDF = inputDF.copy(deep=True)
+        expectedDF[outputCol] = flattenAndConvertNumpy(expectedOutput)
+
+        inputDF = self.spark.createDataFrame(inputDF)
+        actualDF = km.transform(inputDF).toPandas()
+
+        _assertPandasAlmostEqual(actualDF, expectedDF, keyCols + ["useless"])
+
+    NDIM = 5
```

Comment:
RESPONSE REQUESTED: "all these definitions" = are you expecting something other than `NDIM` moved up?